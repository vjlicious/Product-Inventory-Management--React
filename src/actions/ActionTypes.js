export const LOAD_PRODUCTS_SUCCESS = 'LOAD_PRODUCTS_SUCCESS';
export const ADD_PRODUCTS_SUCCESS = 'ADD_PRODUCTS_SUCCESS';
export const DELETE_PRODUCTS_SUCCESS = 'DELETE_PRODUCTS_SUCCESS';
