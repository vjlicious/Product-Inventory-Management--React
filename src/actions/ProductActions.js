import * as types from "./ActionTypes";
import ProductApi from "../data/ProductApi";
import axios from "axios";
export function loadProductsSuccess(products) {
  return { type: types.LOAD_PRODUCTS_SUCCESS, products };
}

export function addProductSuccess(product) {
  return { type: types.ADD_PRODUCTS_SUCCESS, product };
}

export function deleteProductSuccess(id) {
  return { type: types.DELETE_PRODUCTS_SUCCESS, id };
}

export function loadProducts() {
  return function(dispatch) {
    return ProductApi.getAllProducts()
      .then(products => {
        console.log(
          "hi im in load products and sending the data to reducser",
          products
        );
        dispatch(loadProductsSuccess(products));
      })
      .catch(error => {
        throw error;
      });
  };
}
export function getUserDetails() {
  return function(dispatch) {
    return axios
      .get("http://localhost:4000/Users")
      .then(response => {
        dispatch({ type: "GET_USER", payload: response.data });
      })
      .catch();
  };
}

export function incrementCount(product) {
  return function(dispatch) {
    return axios
    .request({
      method: "put",
      url: `http://localhost:4000/productDetails/${product.id}`,
      data: product
    })
    .then(response => {
      dispatch({
        type: "UPDATE_PRODUCT",
        payload: response.data,
        id:product.id
      });
    })
    .catch(err => console.log(err));
  };
}


export function deleteProduct(id) {
  return function(dispatch, getState) {
    return ProductApi.deleteProduct(id)
      .then(response => {
        dispatch(deleteProductSuccess(id));
      })
      .catch(error => {
        throw error;
      });
  };
}
export function getAllData() {
  return function(dispatch) {
    return axios
      .get("http://localhost:4000/db")
      .then(response => {
        dispatch({ type: "GET_ALL", payload: response.data });
      })
      .catch();
  };
}
export function updateProduct(product) {
  return function(dispatch) {
    return axios
      .request({
        method: "put",
        url: `http://localhost:4000/productDetails/${product.id}`,
        data: product
      })
      .then(response => {
        dispatch({
          type: "UPDATE_PRODUCT",
          payload: response.data,
          id: product.id
        });
      })
      .catch(err => console.log(err));
  };
}

export function addProduct(product) {
  return function(dispatch, getState) {
    return ProductApi.saveProduct(product)
      .then(product => {
        dispatch(addProductSuccess(product));
      })
      .catch(error => {
        throw error;
      });
  };
}
