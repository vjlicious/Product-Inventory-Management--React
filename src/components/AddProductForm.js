import React from "react";
import * as ProductActions from "../actions/ProductActions";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { Container, Row, Col, Card } from "react-bootstrap";
import { makeStyles } from "@material-ui/core/styles";
import TextField from "@material-ui/core/TextField";
import ArrowForwardIosIcon from "@material-ui/icons/ArrowForwardIos";
import ArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import Icon from "@material-ui/core/Icon";
import { Button, IconButton } from "@material-ui/core";
import ArrowBackIcon from "@material-ui/icons/ArrowBack";
import { BrowserRouter as Router, Link } from "react-router-dom";
import { notification } from "antd";
import { SmileOutlined } from "@ant-design/icons";

class AddProductForm extends React.Component {
  constructor(props) {
    super(props);
    this.onAddClick = this.onAddClick.bind(this);

    this.handleChange = this.handleChange.bind(this);
    this.state = {
      product: {
        Name: "",
        Price: "",
        Quantity: "",
        Description: "",
        Manufacturer: "",
        count: 0
      },
      Product_error_text: "",
      Price_error: false,
      Price_error_text: "",
      Quantity_error: false,
      Quantity_error_text: "",
      Description_error: false,
      Description_error_text: "",
      Manufacturer_error: false,
      Manufacturer_error_text: ""
    };
  }
  handleChange(event, newValue) {
    event.persist(); // allow native event access (see: https://facebook.github.io/react/docs/events.html)
    // give react a function to set the state asynchronously.
    // here it's using the "name" value set on the TextField
    // to set state.person.[firstname|lastname].

    if (event.target.name == "Price") {
      const num = parseInt(event.target.value);
      if (num <= 0) {
        this.setState({
          Price_error_text: "Enter a valid Price greater than 0"
        });
        this.setState({
          Price_error: true
        });
      } else {
        this.setState({
          Price_error_text: ""
        });
        this.setState({
          Price_error: false
        });
      }
    }
    if (event.target.name == "Quantity") {
      const num = parseInt(event.target.value);
      if (num <= 0) {
        this.setState({
          Quantity_error_text: "Enter a valid Quantity greater than 0"
        });
        this.setState({
          Quantity_error: true
        });
      } else {
        this.setState({
          Quantity_error_text: ""
        });
        this.setState({
          Quantity_error: false
        });
      }
    }
    if (event.target.name == "Description") {
      if (event.target.value.length > 10) {
        this.setState({
          Description_error_text: "Only less than 10 characters allowed"
        });
        this.setState({
          Description_error: true
        });
      } else {
        this.setState({
          Description_error_text: ""
        });
        this.setState({
          Description_error: false
        });
      }
    }
    if (event.target.name == "Manufacturer") {
      if (event.target.value.length > 10) {
        this.setState({
          Manufacturer_error_text: "Only less than 10 characters allowed"
        });
        this.setState({
          Manufacturer_error: true
        });
      } else {
        this.setState({
          Manufacturer_error_text: ""
        });
        this.setState({
          Manufacturer_error: false
        });
      }
    }
    this.setState(
      state => (state.product[event.target.name] = event.target.value)
    );
  }

  onAddClick() {
    const p = this.state.product;
    this.props.actions
      .addProduct(p)
      .then(() => console.log("Product Added"))
      .catch(error => {
        alert(error);
      });
  }
  componentDidMount() {
    console.log("params", this.props);
  }
  render(props) {
    const queryString = window.location.search;

    const urlParams = new URLSearchParams(queryString);

    const page_type = urlParams.get("page_type");

    console.log("yo", window.location.pathname.slice(12, 16));
    console.log("", window.location.pathname.slice(17));
    return (
      <Container fluid>
        <Row
          style={{
            height: "200px"
          }}
        ></Row>
        <Row>
          <Col></Col>
          <Col>
            <Card
              style={{
                height: "500px",
                width: "500px",
                border: "none"
              }}
            >
              <div>
                <form
                  onSubmit={() => {
                    if (
                      this.state.Price_error ||
                      this.state.Quantity_error ||
                      this.state.Description_error ||
                      this.state.Manufacturer_error
                    ) {
                      window.alert("Please check the fields again");
                    } else {
                      this.onAddClick();
                      alert("product added successfully");
                    }
                  }}
                >
                  <Link
                    to={`/dashboard/${window.location.pathname.slice(
                      12,
                      16
                    )}/${window.location.pathname.slice(17)}`}
                    style={{ textDecoration: "none", color: "white" }}
                  >
                    {" "}
                    <IconButton
                      color="primary"
                      aria-label="GO BACK"
                      component="span"
                    >
                      <ArrowBackIcon />
                    </IconButton>
                  </Link>

                  <h1>Enter Product Details</h1>
                  <TextField
                    id="standard-required"
                    label="Product Name"
                    name="Name"
                    required="true"
                    type="text"
                    variant="outlined"
                    onChange={this.handleChange}
                    style={{ height: "100%", width: "100%" }}
                  />

                  <TextField
                    id="standard-required"
                    label="Product Price"
                    variant="outlined"
                    type="number"
                    name="Price"
                    required="true"
                    error={this.state.Price_error}
                    helperText={this.state.Price_error_text}
                    onChange={this.handleChange}
                    style={{
                      height: "100%",
                      width: "100%",
                      marginTop: "10px",
                      textSecurity: "circle" /* IE/Safari */
                    }}
                  ></TextField>
                  <TextField
                    id="standard-required"
                    label="Product's Quantity"
                    variant="outlined"
                    type="number"
                    name="Quantity"
                    error={this.state.Quantity_error}
                    helperText={this.state.Quantity_error_text}
                    required="true"
                    onChange={this.handleChange}
                    style={{
                      height: "100%",
                      width: "100%",
                      marginTop: "10px",
                      textSecurity: "circle" /* IE/Safari */
                    }}
                  ></TextField>
                  <TextField
                    id="standard-required"
                    label="Product's Description"
                    variant="outlined"
                    type="text"
                    name="Description"
                    onChange={this.handleChange}
                    error={this.state.Description_error}
                    helperText={this.state.Description_error_text}
                    required="true"
                    style={{
                      height: "100%",
                      width: "100%",
                      marginTop: "10px",
                      textSecurity: "circle" /* IE/Safari */
                    }}
                  ></TextField>
                  <TextField
                    id="outlined-basic"
                    label="Manufacturer"
                    variant="outlined"
                    type="text"
                    required="true"
                    error={this.state.Manufacturer_error}
                    helperText={this.state.Manufacturer_error_text}
                    name="Manufacturer"
                    onChange={this.handleChange}
                    style={{
                      height: "100%",
                      width: "100%",
                      marginTop: "10px",
                      textSecurity: "circle" /* IE/Safari */
                    }}
                  ></TextField>

                  <div style={{ textAlign: "center", marginTop: "10px" }}>
                    <Button
                      variant="contained"
                      color="primary"
                      type="submit"
                      endIcon={<ArrowForwardIosIcon />}
                      style={{ marginLeft: "10px" }}
                    >
                      Add Product
                    </Button>
                  </div>
                </form>
              </div>
            </Card>
          </Col>
          <Col></Col>
        </Row>
        <Row>
          <h1>{this.props.products}</h1>
          <h1>{console.log("hi got the data from store", this.props.user)}</h1>
        </Row>
      </Container>
    );
  }
}
// const formik = () => {
//   <Formik
//   render={() => {return(<AddProductForm />)}}

//   validationSchema={validationSchema}
// />
// }

function mapStateToProps(state, ownProps) {
  return {
    products: state.products
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(ProductActions, dispatch)
  };
}
export default connect(mapStateToProps, mapDispatchToProps)(AddProductForm);
