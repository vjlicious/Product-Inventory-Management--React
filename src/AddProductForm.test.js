import React from "react";
import { shallow, mount } from "enzyme";
import configureStore from "./store/ConfigureStore";
import { Provider } from "react-redux";

import ProductList from "./components/ProductsList";
import { Table } from "react-bootstrap";
import AddProductForm from "./components/AddProductForm";
import * as actions from "./actions/ProductActions";
import * as actionsTypes from "./actions/ActionTypes";
// const findByTestAtrr = (component, attr) => {
//     const wrapper = component.find(`[data-test='${attr}']`);
//     return wrapper;
// };
// export const testStore = (initialState) => {
//     const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
//     return createStoreWithMiddleware(rootReducer, initialState);
// };

describe("Test ProductList using full shallow rendering", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProductList />);
  });

  it("Add productlist contains a table", () => {
    expect(wrapper.find("table").length).toEqual(1);
  });
});
describe("Test ProductList using full shallow rendering", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProductList />);
  });

  it(" productlist's table should contain 4 headings", () => {
    expect(wrapper.find("th").length).toEqual(4);
  });
});
describe("Test ProductList using full shallow rendering", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProductList />);
  });

  it(" productlist's table ist heading should be Product ID", () => {
    expect(
      wrapper
        .find("tr")
        .childAt(0)
        .render()
        .text()
    ).toEqual("Product ID");
  });
});
describe("Test ProductList using full shallow rendering", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProductList />);
  });

  it(" productlist's table 2nd heading should be Product Name", () => {
    expect(
      wrapper
        .find("tr")
        .childAt(1)
        .render()
        .text()
    ).toEqual("Product Name");
  });
});
describe("Test ProductList using full shallow rendering", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProductList />);
  });

  it(" productlist's table 3rd heading should be Quantity", () => {
    expect(
      wrapper
        .find("tr")
        .childAt(2)
        .render()
        .text()
    ).toEqual("Quantity");
  });
});
describe("Test ProductList using full shallow rendering", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProductList />);
  });

  it(" productlist's table 4th heading should be Quantity", () => {
    expect(
      wrapper
        .find("tr")
        .childAt(3)
        .render()
        .text()
    ).toEqual("Price");
  });
});
describe("check if adpproduct action is despatches correctly", () => {
  let Product;
  let expectedAction;

  beforeEach(() => {
    Product = {
      Name: "Product_1",
      Price: "40000",
      Quantity: "1",
      Description: "Best flagship ",
      Manufacturer: "xyz",
      id: 1,
      count: 600
    };
    expectedAction = {
      type: "ADD_PRODUCTS_SUCCESS",
      product: {
        Name: "Product_1",
        Price: "40000",
        Quantity: "1",
        Description: "Best flagship ",
        Manufacturer: "xyz",
        id: 1,
        count: 600
      }
    };
  });

  it("add product", () => {
    expect(actions.addProductSuccess(Product)).toEqual(expectedAction);
  });
});

describe("check if loadproduct action is despatches correctly", () => {
  let Product;
  let expectedAction;

  beforeEach(() => {
    Product = {
      Name: "Product_1",
      Price: "40000",
      Quantity: "1",
      Description: "Best flagship ",
      Manufacturer: "xyz",
      id: 1,
      count: 600
    };
    expectedAction = {
      type: "LOAD_PRODUCTS_SUCCESS",
      products: {
        Name: "Product_1",
        Price: "40000",
        Quantity: "1",
        Description: "Best flagship ",
        Manufacturer: "xyz",
        id: 1,
        count: 600
      }
    };
  });

  it("add product", () => {
    expect(actions.loadProductsSuccess(Product)).toEqual(expectedAction);
  });
});
describe("check if delete product action is despatches correctly", () => {
  let Product;
  let expectedAction;

  beforeEach(() => {
    Product = {
      Name: "Product_1",
      Price: "40000",
      Quantity: "1",
      Description: "Best flagship ",
      Manufacturer: "xyz",
      id: 1,
      count: 600
    };
    expectedAction = {
      type: 'LOAD_PRODUCTS_SUCCESS', products: 1
    };
  });

  it("delete product", () => {
    console.log(actions.loadProductsSuccess(Product.id));
    expect(actions.loadProductsSuccess(Product.id)).toEqual(expectedAction);
  });
});
