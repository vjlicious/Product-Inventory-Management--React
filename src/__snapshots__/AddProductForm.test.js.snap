// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test  using full dom rendering Add productform renders correctly 1`] = `undefined`;

exports[`Test AddProductForm using Shallow rendering has 5 text input elements 1`] = `undefined`;

exports[`Test AddProductForm using Shallow rendering renders correctly 1`] = `undefined`;

exports[`Test AddProductForm using full dom rendering Add productform renders correctly 1`] = `undefined`;

exports[`Test AddProductForm using full shallow rendering  productlist's table 4th heading should be Quantity 1`] = `undefined`;

exports[`Test AddProductForm using full shallow rendering  render 1`] = `ShallowWrapper {}`;

exports[`Test ProductList using full dom rendering Add productform renders correctly 1`] = `undefined`;
